dj/construction/dj_query.py

```diff
@@ -0,0 +1,230 @@
+from dj.errors import DJException
+from dj.models.database import Database
+from dj.models.node import NodeRevision, NodeType
+from dj.sql.parsing import ast
+from dj.sql.parsing.backends.exceptions import DJParseException
+from dj.sql.parsing.backends.sqloxide import parse
+
+
+def try_get_dj_node(
+    session: Session,
+    name: str,
+    kinds: Set[NodeType],
+) -> Optional[NodeRevision]:
+    "wraps get dj node to return None if no node is found"
+    try:
+        return get_dj_node(session, name, kinds)
+    except DJException:
+        return None
+
+
+def _resolve_metric_nodes(session, col):
+    """
+    Check if a column is a metric and modify the
+    select accordingly
+    """
+    joins = []
+    col_name = make_name(col.namespace, col.name.name)
+    if metric_node := try_get_dj_node(
+        session,
+        col_name,
+        {NodeType.METRIC},
+    ):
+        # if we found a metric node we need to check where it came from
+        parent_select = cast(ast.Select, col.get_nearest_parent_of_type(ast.Select))
+        if not getattr(
+            parent_select,
+            "_validated",
+            False,
+        ):  # pragma: no cover# we haven't seen this
+            if len(parent_select.from_.tables) != 1 or parent_select.from_.joins:
+                raise DJParseException(
+                    "Any SELECT referencing a Metric must source "
+                    "from a single unaliased Table named `metrics`.",
+                )
+            metrics_ref = parent_select.from_.tables[0]
+            try:
+                metrics_ref_name = make_name(
+                    metrics_ref.namespace,  # type: ignore
+                    metrics_ref.name.name,  # type: ignore
+                )
+            except AttributeError:  # pragma: no cover
+                metrics_ref_name = ""
+            if metrics_ref_name != "metrics":
+                raise DJParseException(
+                    "The name of the table in a Metric query must be `metrics`.",
+                )
+            parent_select.from_ = ast.From(
+                [],
+            )  # clear the FROM to prep it for the actual tables
+            parent_select._validated = True  # pylint: disable=W0212
+
+        # we have a metric from `metrics`
+        metric_name = amenable_name(metric_node.name)
+        metric_select = parse(  # pylint: disable=W0212
+            cast(str, metric_node.query),
+        )._to_select()
+        tables = metric_select.from_.tables + [
+            join.table for join in metric_select.from_.joins
+        ]
+        metric_table_expression = ast.Alias(
+            ast.Name(metric_name),
+            None,
+            metric_select,
+        )
+        # we go through all the dep nodes directly in the metric's FROM
+        # we need to surface the node itself to join potential dims
+        # and to surface the node we need to source all its columns
+        # - in the metric for an implicit join
```

Comment:
nit: Since the loop just calls the `_hoist_metric_source_tables`, these comments should probably just go in that function's doc-string